src/jit/valuenum.cpp

```diff
@@ -1164,14 +1207,48 @@ ValueNum ValueNumStore::VNForMapSelect(ValueNumKind vnk, var_types typ, ValueNum
     return result;
 }
 
-ValueNum ValueNumStore::VNForMapSelectWork(ValueNumKind vnk, var_types typ, ValueNum arg0VN, ValueNum arg1VN, unsigned* pBudget)
+//------------------------------------------------------------------------------
+// VNForMapSelectWork : A method that does the work for VNForMapSelect and may call itself recursively.
+//
+//
+// Arguments:
+//    vnk  -             Value number kind
+//    typ  -             Value type
+//    arg0VN  -          Zeroth argument
+//    arg1VN  -          First argument
+//    pBudget -          Remaining budget for the outer evaluation
+//    pUsedRecursiveVN - Out-parameter that is set to true iff RecursiveVN was returned from this method
+//                       or from a method called during one of recursive invocations.
+//
+// Return Value:
+//    Value number for the result of the evaluation.
+//
+// Notes:
+//    This requires a "ValueNumKind" because it will attempt, given "select(phi(m1, ..., mk), ind)", to evaluate
+//    "select(m1, ind)", ..., "select(mk, ind)" to see if they agree.  It needs to know which kind of value number
+//    (liberal/conservative) to read from the SSA def referenced in the phi argument.
+
+ValueNum ValueNumStore::VNForMapSelectWork(ValueNumKind vnk,
+                                           var_types typ,
+                                           ValueNum arg0VN,
+                                           ValueNum arg1VN,
+                                           unsigned* pBudget,
+                                           bool* pUsedRecursiveVN)
 {
     // This label allows us to directly implement a tail call by setting up the arguments, and doing a goto to here.
 TailCall:
+#ifdef DEBUG
+    if (m_pComp->verbose)
+    {
+        printf("    VNForMapSelectWork(" STR_VN "%x, " STR_VN "%x):%s is called\n", arg0VN, arg1VN, varTypeName(typ));
+    }
+#endif
     assert(arg0VN != NoVN && arg1VN != NoVN);
     assert(arg0VN == VNNormVal(arg0VN));  // Arguments carry no exceptions.
     assert(arg1VN == VNNormVal(arg1VN));  // Arguments carry no exceptions.
 
+    *pUsedRecursiveVN = false;
```

Comment:
Does this initialization work properly with regards to the recursive branch to the TailCall label?